docs/proposals/project-scoped-appset.md

```diff
@@ -0,0 +1,114 @@
+
+## Summary
+
+ArgoCD users will be able to use ApplicationSets in the context of the projects where they have access by RBAC. From external usage point of view, each time when an ApplicationSet is created/updated, the controller will validate if the user has permission for doing it. From controller point of view, during the Applications generation, the controller will ensure that the used resources are part of the project 
+
+## Motivation
+
+We have transversal teams where they are the owners of the application in all the stages (code, infra, ux, design,...) and our products are deployed in a cluster per product. Apart from this, critical products are deployed into multiple regions, which means that a product could be in 3-5 different Kubernetes clusters (including different cloud providers). 
+In this use case, ApplicationSet is the perfect match for us because they can deploy an appset for each microservice and automatically they are deployed into all the clusters with the required configuration (also enabling disaster recovery scenarios in other providers/regions thanks to the generators)
+
+As a platform team, we provide to the teams the tooling like ArgoCD, Harbor, observability,... This means that we manage an ArgoCD installation shared for all the teams, managing their clusters from the same place (based on RBAC). We'd like to allow the usage of appsets for the teams, but due to the global scope of them, teams can deploy things into other teams clusters because appsets ignores any kind of limitation.
+
+### Goals
+
+Add support for deploying ApplicationSets inside a AppProject and limit the interaction with elements inside it. 
+This includes also the RBAC verification
+
+### Non-Goals
+
+This doesn't include any change in the UI/cli to make them user friendly
+
+## Proposal
+
+The proposal for achieving this goal is to add a "required" `project` field (read more below) into the ApplicationSet and based on it, limit the access to the resources during the Applications generation. 
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: cluster-addons
+  namespace: argocd
+spec:
+  project: sample-project # new field
+  generators:
+  - ...
+```
+
+As we want to maintain the option of deploying global ApplicationSet (without any restriction), administrators will be able to deploy appsets without setting the project field. For achieving this, we will check if the user has access to **ALL** `applicationsets` across **ALL** the projects (*/*) and only that case, they will be able to deploy `applicationsets` without setting the project:
+
+```
+p, role:admin, applicationsets, create, */*, allow
+p, role:admin, applicationsets, update, */*, allow
+```
+
+Other option is to be explicit and create an specific permission for this:
+
+```
+p, role:admin, applicationsets, unscoped, *, allow
```

Comment:
I think this might be the way to go. afaik, our RBAC library (casbin) does not provide a way to determine whether `*/*` is granted - you ask it "is project X" allowed, and it evaluates the pattern. afaik it has no way to say "all projects are allowed".

Reply:
Nice to know it, in that case I believe that the new role is the better way

Reply:
Argo CD already has project-scoped repositories and clusters. It uses [implicit](https://github.com/argoproj/argo-cd/blob/f61f94f7fc149460e451bf2f698d980025f13e38/docs/user-guide/projects.md?plain=1#L272) RBAC. API server does the trick. If repo/cluster has cluster then API check permissions for `<project>/<name>` otherwise just `<name>`. 

https://github.com/argoproj/argo-cd/blob/cb653f184873e8f3b05547366bae007c7cbf78c0/server/repository/repository.go#L80-L85

The idea is that all admins get:

```
p, role:admin, repositories, create, *, allow
```

and project scoped admins get:

```
p, role:admin, repositories, create, <project-name>/*, allow
```

Any chance we can use the same approach? Proposing it mainly to be consistent with existing features.